src/storage/recovery/recovery_manager.cpp

```diff
@@ -77,19 +79,44 @@ void RecoveryManager::ProcessCommittedTransaction(terrier::transaction::timestam
   // Begin a txn to replay changes with.
   auto *txn = txn_manager_->BeginTransaction();
 
+  // For recording recovery throughput metric
+  uint64_t elapsed_us = 0;
+  uint64_t total_elapsed_us = 0;
+  uint64_t num_txns = 0;
+  uint64_t num_bytes = 0;
+
   // Apply all buffered changes. They should all succeed. After applying we can safely delete the record
   for (uint32_t idx = 0; idx < buffered_changes_map_[txn_id].size(); idx++) {
     auto *buffered_record = buffered_changes_map_[txn_id][idx].first;
     TERRIER_ASSERT(
         buffered_record->RecordType() == LogRecordType::REDO || buffered_record->RecordType() == LogRecordType::DELETE,
         "Buffered record must be a redo or delete.");
+    {
+      common::ScopedTimer<std::chrono::microseconds> scoped_timer(&elapsed_us);
+      if (IsSpecialCaseCatalogRecord(buffered_record)) {
+        idx += ProcessSpecialCaseCatalogRecord(txn, &buffered_changes_map_[txn_id], idx);
+      } else if (buffered_record->RecordType() == LogRecordType::REDO) {
+        ReplayRedoRecord(txn, buffered_record);
+      } else {
+        ReplayDeleteRecord(txn, buffered_record);
+      }
+    }
 
-    if (IsSpecialCaseCatalogRecord(buffered_record)) {
-      idx += ProcessSpecialCaseCatalogRecord(txn, &buffered_changes_map_[txn_id], idx);
-    } else if (buffered_record->RecordType() == LogRecordType::REDO) {
-      ReplayRedoRecord(txn, buffered_record);
-    } else {
-      ReplayDeleteRecord(txn, buffered_record);
+    // Update metrics
+    total_elapsed_us += elapsed_us;
+    num_txns += 1;
+    num_bytes += buffered_record->Size();
+
+    // Check if we need to export throughput metric
+    if (total_elapsed_us >= recovery_metric_interval_) {
+      // Export throughput metric
+      if (num_txns > 0 && common::thread_context.metrics_store_ != nullptr &&
+          common::thread_context.metrics_store_->ComponentEnabled(metrics::MetricsComponent::RECOVERY)) {
+        common::thread_context.metrics_store_->RecordRecoveryData(num_txns, num_bytes);
+        num_txns = 0;
+        num_bytes = 0;
+      }
+      total_elapsed_us = 0;
```

Comment:
Why are you reseting the time period here? If the conditions to export a metric are not met, we should extend the period (by not reseting it) until they are.